core/src/labels/label.cpp

```diff
@@ -21,107 +23,80 @@ Label::Label(Label::Transform _transform, glm::vec2 _size, Type _type, Options _
-    m_dirty = true;
-    m_proxy = false;
-    m_xAxis = glm::vec2(1.0, 0.0);
-    m_yAxis = glm::vec2(0.0, 1.0);
     m_occluded = false;
     m_parent = nullptr;
 }
 
 Label::~Label() {}
 
-void Label::setProxy(bool _proxy) {
-    m_proxy = _proxy;
-}
-
 bool Label::updateScreenTransform(const glm::mat4& _mvp, const glm::vec2& _screenSize, bool _testVisibility) {
 
     glm::vec2 screenPosition;
-    float rot = 0;
-
-    glm::vec2 ap1, ap2;
+    glm::vec2 rotation = {1, 0};
+    bool clipped = false;
 
     switch (m_type) {
         case Type::debug:
         case Type::point:
         {
-            glm::vec4 v1 = worldToClipSpace(_mvp, glm::vec4(m_transform.modelPosition1, 0.0, 1.0));
+            glm::vec2 p0 = m_transform.modelPosition1;
 
-            if (_testVisibility && (v1.w <= 0)) {
+            screenPosition = worldToScreenSpace(_mvp, glm::vec4(p0, 0.0, 1.0),
+                                                _screenSize, clipped);
+
+            if (_testVisibility && clipped) {
                 return false;
             }
 
-            screenPosition = clipToScreenSpace(v1, _screenSize);
-
-            ap1 = ap2 = screenPosition;
+            screenPosition += m_anchor;
 
             break;
         }
         case Type::line:
         {
-            // project label position from mercator world space to clip
+            // project label position from mercator world space to screen
             // coordinates
-            glm::vec4 v1 = worldToClipSpace(_mvp, glm::vec4(m_transform.modelPosition1, 0.0, 1.0));
-            glm::vec4 v2 = worldToClipSpace(_mvp, glm::vec4(m_transform.modelPosition2, 0.0, 1.0));
+            glm::vec2 p0 = m_transform.modelPosition1;
+            glm::vec2 p2 = m_transform.modelPosition2;
+
+            glm::vec2 ap0 = worldToScreenSpace(_mvp, glm::vec4(p0, 0.0, 1.0),
+                                               _screenSize, clipped);
+            glm::vec2 ap2 = worldToScreenSpace(_mvp, glm::vec4(p2, 0.0, 1.0),
+                                               _screenSize, clipped);
 
             // check whether the label is behind the camera using the
             // perspective division factor
-            if (_testVisibility && (v1.w <= 0 || v2.w <= 0)) {
+            if (_testVisibility && clipped) {
                 return false;
             }
 
-            // project to screen space
-            glm::vec2 p1 = clipToScreenSpace(v1, _screenSize);
-            glm::vec2 p2 = clipToScreenSpace(v2, _screenSize);
+            float length = glm::length(ap2 - ap0);
 
-            rot = angleBetweenPoints(p1, p2) + M_PI_2;
+            // default heuristic : allow label to be 30% wider than segment
+            float minLength = m_dim.x * 0.7;
 
-            if (rot > M_PI_2 || rot < -M_PI_2) { // un-readable labels
-                rot += M_PI;
-            } else {
-                std::swap(p1, p2);
+            if (_testVisibility && length < minLength) {
```

Comment:
Ohh I just mean, to do `if (length < minLength)`. `_testVisibility` must be `true` here before of `if (_testVisibility && clipped)` check before.